### PR TITLE
[ipc] Mailbox Examples

### PR DESCRIPTION
**Pull request by hansb21**
In this PR, I implemented 3 examples using mailboxes on nanvix as follows:
- Broadcast: A sends 1 message to B
- Gather: A receives 1 message from B
- Ping pong: A sends and then receives 1 messages from B

### CODE REVIEW

File: src/mailbox/broadcast/main.c [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * @brief Receives messages from worker.
+ */
+static void do_leader(void) 
+{
+	int outboxes;
+	/* Establish connections. */
+	uassert((outboxes= kmailbox_open(PROCESSOR_NODENUM_LEADER + 1, PORT_NUM)) >= 0);
+	
+	uassert(barrier_wait(barrier) == 0);
```
Comment (ppenna) [inline src/mailbox/broadcast/main.c:53]: Why do we need a barrier here? Leave a comment to explain why this is mandatory ;-)

File: src/mailbox/broadcast/main.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+ */
+static void do_worker(void)
+{
+	int inbox;
+	uint64_t latency; 
+	size_t volume;
+
+	uassert((inbox = kmailbox_create(knode_get_num(), PORT_NUM)) >= 0);
+
+	uassert(barrier_wait(barrier) == 0);
```
Comment (ppenna) [inline src/mailbox/broadcast/main.c:78]: Why do we need a barrier here? Leave a comment to explain why this is mandatory ;-)

File: src/mailbox/broadcast/main.c [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <nanvix/runtime/runtime.h>
+#include <nanvix/runtime/barrier.h>
```
Comment (ppenna) [inline src/mailbox/broadcast/main.c:25]: Could you double check if `#include <nanvix/runtime/runtime.h>` includes `#include <nanvix/runtime/barrier.h>`. If yes, just drop the duplicate include here to keep the namespace clean ;-)

File: src/mailbox/gather/main.c [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <nanvix/runtime/runtime.h>
+#include <nanvix/runtime/barrier.h>
```
Comment (ppenna) [inline src/mailbox/gather/main.c:25]: Could you double check if `#include <nanvix/runtime/runtime.h>` includes `#include <nanvix/runtime/barrier.h>`. If yes, just drop the duplicate include here to keep the namespace clean ;-)

File: src/mailbox/gather/main.c [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+ */
+static void do_leader(void)
+{
+	int inbox;
+	uint64_t latency; 
+	size_t volume;
+
+	uassert((inbox = kmailbox_create(knode_get_num(), PORT_NUM)) >= 0);
+
+	uassert(barrier_wait(barrier) == 0);
```
Comment (ppenna) [inline src/mailbox/gather/main.c:55]: Why do we need a barrier here? Leave a comment to explain why this is mandatory ;-)

File: src/mailbox/gather/main.c [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+			KMAILBOX_MESSAGE_SIZE
+		) == KMAILBOX_MESSAGE_SIZE
+	);
+
+	uassert(kmailbox_ioctl(inbox, KMAILBOX_IOCTL_GET_LATENCY, &latency) == 0);
+	uassert(kmailbox_ioctl(inbox, KMAILBOX_IOCTL_GET_VOLUME, &volume) == 0);
+
+	uprintf("[mail][broadcast] latency = %l volume = %d", latency, volume);
+
+	uassert(barrier_wait(barrier) == 0);
```
Comment (ppenna) [inline src/mailbox/gather/main.c:70]: @joaovicentesouto do we need a barrier afterwards also?

File: src/mailbox/gather/main.c [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+ * @brief Receives messages from worker.
+ */
+static void do_worker(void) 
+{
+	int outbox;
+
+	/* Establish connections. */
+	uassert((outbox = kmailbox_open(PROCESSOR_NODENUM_LEADER, PORT_NUM)) >= 0);
+	
+	uassert(barrier_wait(barrier) == 0);
```
Comment (ppenna) [inline src/mailbox/gather/main.c:86]: Why do we need a barrier here? Leave a comment to explain why this is mandatory ;-)

File: src/mailbox/pingpong/main.c [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+*/
+static void do_leader(void) 
+{
+	int inbox, outbox;
+	uint64_t latency, volume;
+
+	uassert((inbox = kmailbox_create(knode_get_num(), PORT_NUM)) >= 0);
+	uassert((outbox = kmailbox_open(PROCESSOR_NODENUM_LEADER + 1, PORT_NUM)) >= 0);
+
+	uassert(barrier_wait(barrier) == 0);
```
Comment (ppenna) [inline src/mailbox/pingpong/main.c:55]: Why do we need a barrier here? Leave a comment to explain why this is mandatory ;-)

File: src/mailbox/pingpong/main.c [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <nanvix/runtime/runtime.h>
+#include <nanvix/runtime/barrier.h>
```
Comment (ppenna) [inline src/mailbox/pingpong/main.c:25]: Could you double check if `#include <nanvix/runtime/runtime.h>` includes `#include <nanvix/runtime/barrier.h>`. If yes, just drop the duplicate include here to keep the namespace clean ;-)

File: src/mailbox/broadcast/main.c [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+static void do_leader(void) 
+{
+	int outboxes;
+	/* Establish connections. */
+	uassert((outboxes= kmailbox_open(PROCESSOR_NODENUM_LEADER + 1, PORT_NUM)) >= 0);
+	
+	uassert(barrier_wait(barrier) == 0);
+
+	/* Broadcast message. */
+	uassert(
```
Comment (joaovicentesouto) [inline src/mailbox/broadcast/main.c:56]: Fix tab indentation to look like this:
```c
	/* Broadcast message. */
	uassert(
		kmailbox_write(
			outbox,
			msg,
			KMAILBOX_MESSAGE_SIZE
		) == KMAILBOX_MESSAGE_SIZE
	);
```

File: src/mailbox/broadcast/main.c [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+{
+	int inbox;
+	uint64_t latency; 
+	size_t volume;
+
+	uassert((inbox = kmailbox_create(knode_get_num(), PORT_NUM)) >= 0);
+
+	uassert(barrier_wait(barrier) == 0);
+
+		uassert(
```
Comment (joaovicentesouto) [inline src/mailbox/broadcast/main.c:80]: Fix tab indentation to look like this:
```c
	uassert(
		kmailbox_read(
			inbox,
			msg,
			KMAILBOX_MESSAGE_SIZE
		) == KMAILBOX_MESSAGE_SIZE
	);
```

File: src/mailbox/gather/main.c (hunk 6 above)
Comment (joaovicentesouto) [reply src/mailbox/gather/main.c:70]: No, we only need one after the `create`/`open` to ensure that another thread does not drop any message destined to user mailboxes ([microkernel issue](https://github.com/nanvix/microkernel/issues/268)).
@hansb21, this issue is why we need the barrier. And so, we can remove all secondary `barrier_wait`.

File: src/mailbox/gather/main.c [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	for (int i = 0; i < NCLUSTERS; i++)
+		nodes[i] = PROCESSOR_NODENUM_LEADER + i;
+
+	barrier = barrier_create(nodes, NCLUSTERS);
+	uassert(BARRIER_IS_VALID(barrier));
+
+		fn();
+
+	uassert(barrier_destroy(barrier) == 0);
+}
```
Comment (joaovicentesouto) [inline src/mailbox/gather/main.c:121]: Need a new line between function.

File: src/mailbox/pingpong/main.c [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+	uassert(barrier_wait(barrier) == 0);
+
+	uassert(kmailbox_write(outbox, msg, KMAILBOX_MESSAGE_SIZE) == KMAILBOX_MESSAGE_SIZE);
+	uassert(kmailbox_read(inbox, msg, KMAILBOX_MESSAGE_SIZE) == KMAILBOX_MESSAGE_SIZE);
+
+	uassert(barrier_wait(barrier) == 0);
+
+	uassert(kmailbox_close(outbox) == 0);
+	uassert(kmailbox_unlink(inbox) == 0);
+}
```
Comment (joaovicentesouto) [inline src/mailbox/pingpong/main.c:90]: Need a new line between function.

File: src/mailbox/broadcast/main.c (hunk 3 above)
Comment (hansb21) [reply src/mailbox/broadcast/main.c:25]: `#include <nanvix/runtime/runtime.h>` dosen't includes `#include <nanvix/runtime/barrier.h>`.

File: src/mailbox/gather/main.c (hunk 4 above)
Comment (hansb21) [reply src/mailbox/gather/main.c:25]: `#include <nanvix/runtime/runtime.h> ` dosen't includes `#include <nanvix/runtime/barrier.h>`.

File: src/mailbox/pingpong/main.c (hunk 9 above)
Comment (hansb21) [reply src/mailbox/pingpong/main.c:25]: `#include <nanvix/runtime/runtime.h>` dosen't includes `#include <nanvix/runtime/barrier.h>`.

File: src/mailbox/pingpong/main.c (hunk 8 above)
Comment (hansb21) [reply src/mailbox/pingpong/main.c:55]: We need a barrier here to ensure that another thread does not drop any message destined to user mailboxes. [microkernel issue](https://github.com/nanvix/microkernel/issues/268)

File: src/mailbox/gather/main.c (hunk 7 above)
Comment (hansb21) [reply src/mailbox/gather/main.c:86]: We need a barrier here to ensure that another thread does not drop any message destined to user mailboxes. [(microkernel issue)](https://github.com/nanvix/microkernel/issues/268)

File: src/mailbox/gather/main.c (hunk 5 above)
Comment (hansb21) [reply src/mailbox/gather/main.c:55]: We need a barrier here to ensure that another thread does not drop any message destined to user mailboxes. [(microkernel issue)](https://github.com/nanvix/microkernel/issues/268)

File: src/mailbox/broadcast/main.c (hunk 2 above)
Comment (hansb21) [reply src/mailbox/broadcast/main.c:78]: We need a barrier here to ensure that another thread does not drop any message destined to user mailboxes. [(microkernel issue)](https://github.com/nanvix/microkernel/issues/268)

File: src/mailbox/broadcast/main.c (hunk 1 above)
Comment (hansb21) [reply src/mailbox/broadcast/main.c:53]: We need a barrier here to ensure that another thread does not drop any message destined to user mailboxes. [(microkernel issue)](https://github.com/nanvix/microkernel/issues/268)

File: src/mailbox/broadcast/main.c (hunk 2 above)
Comment (ppenna) [reply src/mailbox/broadcast/main.c:77]: @hansb21 Yes =) could you add this comment to the source code?

**Pull request merged**